### PR TITLE
Proposal: add a runtime pod cache in kubelet

### PR DESCRIPTION
**Pull request by yujuhong**

### CODE REVIEW

File: docs/proposals/runtime-pod-cache.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+Runtime pod cache is an in-memory cache which stores the *states* of
+all pods, and is maintained by PLEG. It serves as a single source of
+truth for pod states, freeing Kubelet from querying the container
+runtime for states.
+
+## Motivation
+
+With PLEG, Kubelet no longer needs to perform comprehensive state
+checking for all pods periodically. It only instructs a pod worker to
+start syncing when there is a change of its pod state. Nevertheless,
```
Comment (dchen1107) [inline docs/proposals/runtime-pod-cache.md]: Changing "there is a change of its pod state" to something like "there is a change of container state within the pod"? For a pod with RestartPolicy:Always, the pod state stays at Running until someone destroys it.

**Comment by yujuhong** [comment]
@dchen1107, I updated the proposal. PTAL, thanks!

**Comment by eparis** [comment]
@k8s-bot test this issue: #IGNORE

Tests have been pending for 24 hours

File: docs/proposals/runtime-pod-cache.md [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    Image string
+    ImageID string
+    Hash uint64
+    RestartCount int
+    Reason string
+    Message string
+}
+```
+
+`PodStatus` is defined in the container runtime interface, hence is
```
Comment (dchen1107) [inline docs/proposals/runtime-pod-cache.md]: s/is is /is ?

File: docs/proposals/runtime-pod-cache.md [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+## TODOs for v1.2
+
+ - Redefine container runtime types ([#12619](https://issues.k8s.io/12619)):
+   and introduce `PodStatus`. Refactor dockertools and rkt to use the new type.
+
+ - Add cache and instruct PLEG to populate it.
+
+ - Refactor Kubelet to use the cache.
+
+ - Deprecate the old runtiem cache.
```
Comment (dchen1107) [inline docs/proposals/runtime-pod-cache.md]: s/runtiem/runtime?

**Comment by dchen1107** [comment]
Small nits. Otherwise lgtm

File: docs/proposals/runtime-pod-cache.md (hunk 3 above)
Comment (yujuhong) [inline docs/proposals/runtime-pod-cache.md]: done

File: docs/proposals/runtime-pod-cache.md (hunk 2 above)
Comment (yujuhong) [inline docs/proposals/runtime-pod-cache.md]: done

**Comment by yujuhong** [comment]
Fixed the typos. Thanks!

**Comment by yujuhong** [comment]
@dchen1107, ping! Thanks.

**Comment by dchen1107** [comment]
LGTM

**Comment by dchen1107** [comment]
LGTM

**Pull request merged**